src/model/warning.ts

```diff
@@ -14,24 +14,40 @@
 
 import * as chalk from 'chalk';
 
+import {Analyzer} from '../core/analyzer';
 import {ParsedDocument} from '../parser/document';
 import {underlineCode} from '../warning/code-printer';
 
-import {SourceRange} from './source-range';
+import {Analysis} from './analysis';
+import {Document} from './document';
+import {comparePositionAndRange, isPositionInsideRange, SourceRange} from './source-range';
+
+import stable = require('stable');
 
 export interface WarningInit {
   readonly message: string;
   readonly sourceRange: SourceRange;
   readonly severity: Severity;
   readonly code: string;
   readonly parsedDocument: ParsedDocument;
+  readonly fix?: Edit;
+  readonly actions?: ReadonlyArray<Action>;
 }
 export class Warning {
   readonly code: string;
   readonly message: string;
   readonly sourceRange: SourceRange;
   readonly severity: Severity;
 
+  /**
+   * If the problem has a single automatic fix, this is it.
+   *
+   * Whether and how much something is 'automatic' can be a bit tricky to
+   * delineate. Roughly speaking, if 99% of the time the change solves the
+   * issue completely then it should go in `fix`.
+   */
+  readonly fix: Edit|undefined;
```

Comment:
why not `fix?`?

Reply:
I was thinking about this style for consistent object shape. It doesn't actually assert that we assign to the property in the constructor though so it doesn't actually matter though.

Happy to change to `fix?`, I guess I'm just holding a torch for https://github.com/Microsoft/TypeScript/issues/8476

Reply:
without an initializer the emitted code is the same, and I find `?` easier to read by a hair. If you care about share you need add `= undefined`